package android_sensorfirewall;

// !!! ----------------------------------------------------------------------
// !!! Make sure to make identical change in
// !!! frameworks/base/services/java/com/android/server/FirewallConfigMessages.proto
// !!! ----------------------------------------------------------------------

option optimize_for = LITE_RUNTIME;

message Perturb {
    optional float mean = 1; // used for Gaussian
    optional float variance = 2; // used for Gaussian
    optional float unifMin = 3; // lower value of range for uniform
    optional float unifMax = 4; // upper value of range for uniform
    optional float expParam = 5; // parameter for exponential
    enum DistributionType {
        GAUSSIAN = 1;
        UNIFORM = 2;
        EXPONENTIAL = 3;
    }
    required DistributionType distType = 6;
}

message VectorValue {
    optional float x = 1; // x-axis value
    optional float y = 2; // y-axis value
    optional float z = 3; // z-axis value
    optional float theta = 4; // for type-rotation
    optional float accuracy = 5; // for type-rotation
    optional double lat = 6; // gps latitude
    optional double lon = 7; // gps longitude
    optional float speed = 8; // gps speed
    optional float bearing = 9; // gps bearing
}

message SensorValue {
    optional float scalarVal = 1; // for one axis sensors e.g., light
    optional VectorValue vecVal = 2; // used for sensors with multiple axis e.g., acc, magnetic 
    optional int64 timestamp = 3; // not used as yet
    optional float defaultVal = 4; // if vector or scalar value is not set default is used
}

message Param {
    optional SensorValue constantValue = 1; // refers to sensorvalue msg
    optional float delay = 2; // not used
    optional Perturb perturb = 3; // refers to the perturb msg
}

message DateTime {
    repeated uint32 dayOfWeek = 1; // 0-6 with Sunday being 0
    optional uint32 fromHr = 2; // start hour field in 24-hr format
    optional uint32 fromMin = 3; // start minute field
    optional uint32 toHr = 4; // end hour field
    optional uint32 toMin = 5; // end minute field
}

message Action {
    enum ActionType {
        ACTION_SUPPRESS = 1;
        ACTION_CONSTANT = 2;
        ACTION_PLAYBACK = 3;
        ACTION_PERTURB = 4;
        ACTION_PASSTHROUGH = 5;
    }
    required ActionType actionType = 1;
    optional Param param = 2;
}

message Rule {
    required string ruleName = 1;
    required int32 sensorType = 2;
    required string pkgName = 3;
    required uint32 pkgUid = 4;
    optional DateTime dateTime = 5;
    required Action action = 6;
}

message FirewallConfig {
    repeated Rule rule = 1;
}
